### PR TITLE
HSM

### PR DESCRIPTION
**Pull request by fmorency**
First draft.

Add HSM support to MANY client.

CICD will probably fail - I need to fix it to support HSM tests.

Design documentation as well as HSM Primer are on Coda.

Tested against SoftHSMv2 and Yubikey NFC 5

### CODE REVIEW

File: src/many-cli/src/main.rs [hunk 1]
```diff
@@ -159,10 +182,41 @@ fn main() {
             }
         }
         SubCommand::Message(o) => {
-            // If `pem` is not provided, use anonymous and don't sign.
-            let key = o.pem.map_or_else(CoseKeyIdentity::anonymous, |p| {
-                CoseKeyIdentity::from_pem(&std::fs::read_to_string(&p).unwrap()).unwrap()
-            });
+            let key = if o.module.is_some() && o.slot.is_some() && o.keyid.is_some() {
```
Comment (hansl) [inline src/many-cli/src/main.rs:185]:
```suggestion
            let key = if let (Some(module), Some(slot), Some(keyid)) = (o.module, o.slot, o.keyid) {
```

File: src/many-cli/src/main.rs [hunk 2]
```diff
@@ -159,10 +182,41 @@ fn main() {
...
+                trace!("Getting user PIN");
+                let pin = rpassword::prompt_password("Please enter the HSM user PIN: ").unwrap();
+                let keyid = hex::decode(o.keyid.unwrap()).expect("Failed to decode keyid to hex");
+
+                {
+                    let mut hsm = HSM_INSTANCE.lock().unwrap();
+                    hsm.init(o.module.unwrap(), keyid)
```
Comment (hansl) [inline src/many-cli/src/main.rs:192]:
```suggestion
                    hsm.init(module, keyid)
```

File: src/many-cli/src/main.rs [hunk 3]
```diff
@@ -159,10 +182,41 @@ fn main() {
...
-            let key = o.pem.map_or_else(CoseKeyIdentity::anonymous, |p| {
-                CoseKeyIdentity::from_pem(&std::fs::read_to_string(&p).unwrap()).unwrap()
-            });
+            let key = if o.module.is_some() && o.slot.is_some() && o.keyid.is_some() {
+                trace!("Getting user PIN");
+                let pin = rpassword::prompt_password("Please enter the HSM user PIN: ").unwrap();
+                let keyid = hex::decode(o.keyid.unwrap()).expect("Failed to decode keyid to hex");
+
+                {
+                    let mut hsm = HSM_INSTANCE.lock().unwrap();
```
Comment (hansl) [inline src/many-cli/src/main.rs:191]: Can you make it a function of `hsm`? Like

```rust
let mut hsm: MutexGuard<HSM> = HSM::get_intance();
```

File: src/many-cli/src/main.rs [hunk 4]
```diff
@@ -159,10 +182,41 @@ fn main() {
...
+                    hsm.init(o.module.unwrap(), keyid)
+                        .expect("Failed to initialize HSM module");
+
+                    // The session will stay open until the application terminates
+                    hsm.open_session(
+                        o.slot.unwrap(),
```
Comment (hansl) [inline src/many-cli/src/main.rs:197]:
```suggestion
                        slot,
```

File: src/many-cli/src/main.rs [hunk 5]
```diff
@@ -159,10 +182,41 @@ fn main() {
...
         }
         SubCommand::Message(o) => {
-            // If `pem` is not provided, use anonymous and don't sign.
-            let key = o.pem.map_or_else(CoseKeyIdentity::anonymous, |p| {
-                CoseKeyIdentity::from_pem(&std::fs::read_to_string(&p).unwrap()).unwrap()
-            });
+            let key = if o.module.is_some() && o.slot.is_some() && o.keyid.is_some() {
+                trace!("Getting user PIN");
+                let pin = rpassword::prompt_password("Please enter the HSM user PIN: ").unwrap();
+                let keyid = hex::decode(o.keyid.unwrap()).expect("Failed to decode keyid to hex");
```
Comment (hansl) [inline src/many-cli/src/main.rs:188]:
```suggestion
                let keyid = hex::decode(keyid).expect("Failed to decode keyid to hex");
```

File: src/many-cli/src/main.rs [hunk 6]
```diff
@@ -159,10 +182,41 @@ fn main() {
             }
         }
         SubCommand::Message(o) => {
-            // If `pem` is not provided, use anonymous and don't sign.
-            let key = o.pem.map_or_else(CoseKeyIdentity::anonymous, |p| {
-                CoseKeyIdentity::from_pem(&std::fs::read_to_string(&p).unwrap()).unwrap()
-            });
+            let key = if o.module.is_some() && o.slot.is_some() && o.keyid.is_some() {
+                trace!("Getting user PIN");
+                let pin = rpassword::prompt_password("Please enter the HSM user PIN: ").unwrap();
```
Comment (hansl) [inline src/many-cli/src/main.rs:187]: What are you thoughts on non-interactive TTY?

Comment (hansl) [reply src/many-cli/src/main.rs:187]: Please test it. We might just want to disallow non-interactive TTY and tell the user (there's an `atty` crate to get caps of your STDIN)

File: src/many/src/types/identity/cose.rs [hunk 7]
```diff
@@ -56,24 +59,45 @@ impl CoseKeyIdentity {
         Self {
             identity: Identity::anonymous(),
             key: None,
+            hsm: false,
         }
     }
 
-    pub fn from_key(key: CoseKey) -> Result<Self, String> {
+    pub fn from_key(key: CoseKey, hsm: bool) -> Result<Self, String> {
```
Comment (hansl) [inline src/many/src/types/identity/cose.rs:66]: Can we make this private? I don't think people should use this by itself (without using `from_hsm` or `from_pem`).

File: src/many-cli/src/main.rs (hunk 6 above)
Comment (fmorency) [reply src/many-cli/src/main.rs:187]: Right now, doing `echo 1234 | many message ...` blocks on the password prompt, i.e., it doesn't feed the `echo` as the password.

Comment (fmorency) [reply src/many-cli/src/main.rs:187]: Do you want me to add an `atty` check and terminate the process if it's not a TTY?

File: src/many/src/types/identity/cose.rs (hunk 7 above)
Comment (fmorency) [reply src/many/src/types/identity/cose.rs:66]: We're using it in `src/many/src/message.rs`. Would a `pub(crate)` be good enough for you until we refactor?

File: src/many-cli/src/main.rs (hunk 3 above)
Comment (fmorency) [reply src/many-cli/src/main.rs:191]: I wrote it so it returns a `Result` since this thing can panic.

File: src/many-cli/src/main.rs (hunk 6 above)
Comment (hansl) [reply src/many-cli/src/main.rs]: That's perfect, that's what I would expect. We can error out on it but that's probably fine.

File: src/many/src/types/identity/cose.rs (hunk 7 above)
Comment (hansl) [reply src/many/src/types/identity/cose.rs]: Yes.

File: src/many/src/types/identity/cose.rs [hunk 8]
```diff
@@ -226,7 +223,10 @@ impl Signer<CoseKeyIdentitySignature> for CoseKeyIdentity {
                 Algorithm::None => Err(Error::new()),
                 Algorithm::ECDSA => {
                     if self.hsm {
-                        let hsm = HSM_INSTANCE.lock().unwrap();
+                        let hsm = HSM::get_instance().map_err(|e| {
+                            eprintln!("HSM mutex poisoned {}", e);
```
Comment (hansl) [inline src/many/src/types/identity/cose.rs:227]: Please use `error!()` from the tracing module instead of `eprintln!`, here and everywhere else.

**Review by hansl — APPROVED** [review]
One quick change but otherwise LGTM. Pre-emptively approve, but please address the above comment. Good work!

File: src/many/src/types/identity/cose.rs (hunk 8 above)
Comment (fmorency) [reply src/many/src/types/identity/cose.rs]: Done. I took the liberty of removing every `eprintln!` from `many-rs`.

**Review by hansl — APPROVED** [review]

**Pull request merged**